<a href="https://colab.research.google.com/github/w21280485/assignment_a2_2/blob/main/Code_Development_Notebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Load the Dataset

In [1]:
# Import libraries and Load the Dataset

import sqlite3
import pandas as pd

In [2]:
# Establish the connection to the database

connection = sqlite3.connect('cardiohealth.db')
cursor = connection.cursor()

In [3]:
# Inspect the database schema and find out what tables exist

sql='''
SELECT name FROM sqlite_master WHERE type='table';
'''
cursor.execute(sql)
cursor.fetchall()

[('medical_examination',)]

In [5]:
# Load the table into Pandas dataframe

sql='''
SELECT * from medical_examination;
'''
df = pd.read_sql_query(sql, connection)
df

id    age  gender  height  weight  ap_hi  ap_lo  cholesterol  gluc  \
0          0  18393       2     168    62.0    110     80            1     1   
1          1  20228       1     156    85.0    140     90            3     1   
2          2  18857       1     165    64.0    130     70            3     1   
3          3  17623       2     169    82.0    150    100            1     1   
4          4  17474       1     156    56.0    100     60            1     1   
...      ...    ...     ...     ...     ...    ...    ...          ...   ...   
69995  99993  19240       2     168    76.0    120     80            1     1   
69996  99995  22601       1     158   126.0    140     90            2     2   
69997  99996  19066       2     183   105.0    180     90            3     1   
69998  99998  22431       1     163    72.0    135     80            1     2   
69999  99999  20540       1     170    72.0    120     80            2     1   

       smoke  alco  active  cardio  
0          0     0       1       0  
1          0     0       1       1  
2          0     0       0       1  
3          0     0       1       1  
4          0     0       0       0  
...      ...   ...     ...     ...  
69995      1     0       1       0  
69996      0     0       1       1  
69997      0     1       0       1  
69998      0     0       0       1  
69999      0     0       1       0  

[70000 rows x 13 columns]

## Data Exploration

In [15]:
# Display the first few rows of the dataset using head()

df.head(10)

id    age  gender  height  weight  ap_hi  ap_lo  cholesterol  gluc  smoke  \
0   0  18393       2     168    62.0    110     80            1     1      0   
1   1  20228       1     156    85.0    140     90            3     1      0   
2   2  18857       1     165    64.0    130     70            3     1      0   
3   3  17623       2     169    82.0    150    100            1     1      0   
4   4  17474       1     156    56.0    100     60            1     1      0   
5   8  21914       1     151    67.0    120     80            2     2      0   
6   9  22113       1     157    93.0    130     80            3     1      0   
7  12  22584       2     178    95.0    130     90            3     3      0   
8  13  17668       1     158    71.0    110     70            1     1      0   
9  14  19834       1     164    68.0    110     60            1     1      0   

   alco  active  cardio  
0     0       1       0  
1     0       1       1  
2     0       0       1  
3     0       1       1  
4     0       0       0  
5     0       0       0  
6     0       1       0  
7     0       1       1  
8     0       1       0  
9     0       0       0

In [17]:
# Display the bottom few rows of the dataset using tail()

df.tail(10)

id    age  gender  height  weight  ap_hi  ap_lo  cholesterol  gluc  \
69990  99986  15094       1     168    72.0    110     70            1     1   
69991  99988  20609       1     159    72.0    130     90            2     2   
69992  99990  18792       1     161    56.0    170     90            1     1   
69993  99991  19699       1     172    70.0    130     90            1     1   
69994  99992  21074       1     165    80.0    150     80            1     1   
69995  99993  19240       2     168    76.0    120     80            1     1   
69996  99995  22601       1     158   126.0    140     90            2     2   
69997  99996  19066       2     183   105.0    180     90            3     1   
69998  99998  22431       1     163    72.0    135     80            1     2   
69999  99999  20540       1     170    72.0    120     80            2     1   

       smoke  alco  active  cardio  
69990      0     0       1       1  
69991      0     0       1       0  
69992      0     0       1       1  
69993      0     0       1       1  
69994      0     0       1       1  
69995      1     0       1       0  
69996      0     0       1       1  
69997      0     1       0       1  
69998      0     0       0       1  
69999      0     0       1       0

In [7]:
# Get general information about the dataset using info()

df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70000 entries, 0 to 69999
Data columns (total 13 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id           70000 non-null  int64  
 1   age          70000 non-null  int64  
 2   gender       70000 non-null  int64  
 3   height       70000 non-null  int64  
 4   weight       70000 non-null  float64
 5   ap_hi        70000 non-null  int64  
 6   ap_lo        70000 non-null  int64  
 7   cholesterol  70000 non-null  int64  
 8   gluc         70000 non-null  int64  
 9   smoke        70000 non-null  int64  
 10  alco         70000 non-null  int64  
 11  active       70000 non-null  int64  
 12  cardio       70000 non-null  int64  
dtypes: float64(1), int64(12)
memory usage: 6.9 MB


In [8]:
# Summary statistics of numerical columns using describle()

df.describe()

id           age        gender        height        weight  \
count  70000.000000  70000.000000  70000.000000  70000.000000  70000.000000   
mean   49972.419900  19468.865814      1.349843    164.359229     74.205690   
std    28851.302323   2467.251667      0.477253      8.210126     14.395757   
min        0.000000  10798.000000      1.000000     55.000000     10.000000   
25%    25006.750000  17664.000000      1.000000    159.000000     65.000000   
50%    50001.500000  19703.000000      1.000000    165.000000     72.000000   
75%    74889.250000  21327.000000      2.000000    170.000000     82.000000   
max    99999.000000  23713.000000      3.000000    250.000000    200.000000   

              ap_hi         ap_lo   cholesterol          gluc         smoke  \
count  70000.000000  70000.000000  70000.000000  70000.000000  70000.000000   
mean     128.817286     96.630414      1.366871      1.226457      0.088129   
std      154.011419    188.472530      0.680250      0.572270      0.283484   
min     -150.000000    -70.000000      1.000000      1.000000      0.000000   
25%      120.000000     80.000000      1.000000      1.000000      0.000000   
50%      120.000000     80.000000      1.000000      1.000000      0.000000   
75%      140.000000     90.000000      2.000000      1.000000      0.000000   
max    16020.000000  11000.000000      3.000000      3.000000      1.000000   

               alco        active        cardio  
count  70000.000000  70000.000000  70000.000000  
mean       0.053771      0.803729      0.499700  
std        0.225568      0.397179      0.500003  
min        0.000000      0.000000      0.000000  
25%        0.000000      1.000000      0.000000  
50%        0.000000      1.000000      0.000000  
75%        0.000000      1.000000      1.000000  
max        1.000000      1.000000      1.000000

### Discussion

Above statistics indicate that the max value of gender is 3, which is not the same in the assignment description. Handling of unknown dataset is required in the follow section.

## Pre-processing and cleaning data

### Check for unknown values

In [51]:
# Display all the rows with gender = 3 using sql

sql='''
SELECT * from medical_examination WHERE gender = 3
'''
df = pd.read_sql_query(sql, connection)
df

id    age  gender  height  weight  ap_hi  ap_lo  cholesterol  gluc  \
0    2171  18395       3     159    77.0    140    100            1     1   
1    2769  17458       3     155    50.0     90     50            1     1   
2    3898  20376       3     158    58.0    130     80            1     1   
3   13328  23228       3     156    69.0    140     80            3     1   
4   36717  18422       3     158    73.0    110     80            1     1   
5   37141  21977       3     165    80.0    120     80            1     1   
6   37608  19686       3     165    68.0    120     80            1     1   
7   37757  21882       3     163    74.0    150    100            2     1   
8   62308  22536       3     148    86.0    120     80            1     1   
9   89890  19882       3     169    72.0    130     90            1     1   
10  97415  16743       3     161    61.0    120     80            1     1   

    smoke  alco  active  cardio  
0       0     0       1       1  
1       0     0       1       0  
2       0     0       1       0  
3       0     0       1       1  
4       0     0       1       0  
5       0     0       1       0  
6       0     0       1       0  
7       0     0       0       1  
8       0     0       1       0  
9       0     0       1       1  
10      0     0       1       0

### Discussion

Above result indicate that there are 11 rows with gender = 3.
Since we do not have information what the gender = 3 represents, these rows will be removed when data analysis related to gender.
However, if the analysis is not related to gender, all the dataset can stil be used.

### Check for duplicate rows

In [21]:
# Check for duplicates using duplicated()

df.duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
69995    False
69996    False
69997    False
69998    False
69999    False
Length: 70000, dtype: bool

--> Above result indicate that there are no duplicate rows for the first few rows and bottom few rows, but the middle rows cannot be seen. So I try to display the duplicate rows to make sure if no duplicate.

In [23]:
# Try to display duplicate rows

df[df.duplicated()]


Empty DataFrame
Columns: [id, age, gender, height, weight, ap_hi, ap_lo, cholesterol, gluc, smoke, alco, active, cardio]
Index: []

--> Above result indicate that there are no duplicate rows display. I try to display duplicate rows in subset to cross-check the dataset

In [36]:
# Try to display duplicate rows for subsets
df[df.duplicated(subset=["age", "gender", "height", "weight", 'ap_hi', 'ap_lo', 'cholesterol', 'gluc', 'smoke','alco', 'active', 'cardio'])]


id    age  gender  height  weight  ap_hi  ap_lo  cholesterol  gluc  \
10562  15094  20495       1     165    70.0    120     80            1     1   
21784  31110  16793       1     165    68.0    120     80            1     1   
38505  54977  18988       1     164    65.0    120     80            1     1   
40365  57690  14552       1     158    64.0    120     80            1     1   
42450  60642  18353       1     169    67.0    120     80            1     1   
44653  63776  16937       2     170    70.0    120     80            1     1   
45125  64445  21280       1     165    65.0    120     80            1     1   
45748  65352  22077       1     175    69.0    120     80            1     1   
45810  65438  21230       1     164    62.0    120     80            1     1   
48917  69842  21945       1     165    60.0    120     80            1     1   
50432  71971  17493       2     169    74.0    120     80            1     1   
52552  74969  21943       1     165    65.0    120     80            1     1   
56643  80859  17535       2     165    65.0    120     80            1     1   
56906  81232  20293       1     162    70.0    110     70            1     1   
57946  82698  18955       1     165    75.0    120     80            1     1   
58730  83812  19858       1     165    68.0    120     80            1     1   
60453  86312  20516       1     164    66.0    120     80            1     1   
60474  86345  16805       1     157    67.0    120     80            1     1   
62318  88987  18979       1     165    65.0    120     80            1     1   
64169  91592  16160       1     168    65.0    120     80            1     1   
65079  92891  18210       1     160    60.0    120     80            1     1   
65622  93659  21778       1     160    58.0    120     80            1     1   
66190  94486  19059       1     165    65.0    120     80            1     1   
68281  97521  21119       1     160    60.0    120     80            1     1   

       smoke  alco  active  cardio  
10562      0     0       1       0  
21784      0     0       1       0  
38505      0     0       1       0  
40365      0     0       1       0  
42450      0     0       1       0  
44653      0     0       0       0  
45125      0     0       1       0  
45748      0     0       1       1  
45810      0     0       1       0  
48917      0     0       1       0  
50432      0     0       1       1  
52552      0     0       1       1  
56643      0     0       1       0  
56906      0     0       1       0  
57946      0     0       1       1  
58730      0     0       1       0  
60453      0     0       0       0  
60474      0     0       1       0  
62318      0     0       0       0  
64169      0     0       1       1  
65079      0     0       1       0  
65622      0     0       1       0  
66190      0     0       1       1  
68281      0     0       0       1

--> Above result indicate that there are duplicate rows for the subsets excluding 'id', which make sense. I further use sql to call one of the duplicate row age = 20495 to confirm.

In [39]:
sql='''
SELECT * from medical_examination WHERE age = 20495 AND height = 165;
'''
df = pd.read_sql_query(sql, connection)
df

id    age  gender  height  weight  ap_hi  ap_lo  cholesterol  gluc  \
0   3247  20495       1     165    70.0    120     80            1     1   
1  14026  20495       1     165    71.0    120     80            1     1   
2  15094  20495       1     165    70.0    120     80            1     1   

   smoke  alco  active  cardio  
0      0     0       1       0  
1      0     0       1       1  
2      0     0       1       0

### Discussion

Above sql result indicate that id 3247 & id 15094 are having the same data excluding 'id'. This make sense to me that all the ids are unique and there are some individuals having same values in other columns. As such, checking duplicate process is done and no need to remove any duplicate values

### Check for missing values

In [12]:
# Check for missing values using isnull()

df.isnull().sum()

id             0
age            0
gender         0
height         0
weight         0
ap_hi          0
ap_lo          0
cholesterol    0
gluc           0
smoke          0
alco           0
active         0
cardio         0
dtype: int64

In [40]:
# Check for missing values in each row

(df[df['id'].isnull()])

Empty DataFrame
Columns: [id, age, gender, height, weight, ap_hi, ap_lo, cholesterol, gluc, smoke, alco, active, cardio]
Index: []

### Discussion

Above result indicate there are no missing value. As such, checking for missing value is done and no need to handle (remove or impute) missing value.